### PR TITLE
Added test_memory_timeline_no_id

### PR DESCRIPTION
**Pull request by jason-liu227**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [ ] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:

Added test_memory_timeline_no_id to test_spyre_profiler.py

#### Which issue(s) this PR is related to:

This is part of #2448

### CODE REVIEW

**Comment by anubhavjana** [comment]
@prasannabiswas-iitb PTAL -- we might need to update the `test_profiler_config.yaml` if it does not cover this test.

cc: @ashokponkumar @ppnaik1890

**Comment by prasannabiswas-iitb** [comment]
Hi @jason-liu227 , thanks for the PR. 

#### A bit of a background

This test **as it is like in pytorch upstream** used to pass until this PR: https://github.com/torch-spyre/torch-spyre/pull/2163

#### Issue was

We see the ci-cd failure reporting that we are getting size of 8192.

Expected 4096 but got 8192.
Absolute difference: 4096 (up to 0.001 allowed)
Relative difference: 1.0 (up to 0.0 allowed)

```
The failure occurred for item [0][1]
expected does not match actual: [(<Action.PREEXISTING: 1>, 8192), (<Action.DESTROY: 4>, 8192), (<Action.CREATE: 2>, 256), (<Action.DESTROY: 4>, 256), (<Action.CREATE: 2>, 1024), (<Action.DESTROY: 4>, 1024)]
----------------------------- Captured stderr call -----------------------------
/home/senuser/pytorch/test/profiler/test_memory_profiler.py:1568: UserWarning: Backend Spyre does not support int64; downcasting to int32 may change values outside the 32-bit range. You can silence this via warnings.filterwarnings(...) or spyre.set_downcast_warning(False) or TORCH_SPYRE_DOWNCAST_WARN env. variable. (Triggered internally at /home/senuser/torch-spyre/torch_spyre/csrc/types_mapping.h:96.)
  x = torch.ones((1024,), device=device)
```

#### Right now, I still see the assert for (x) 1024 tensor to be 4096. And worried if this will start breaking the PRs again.
Have you tested this locally?

**Comment by ani300** [comment]
The reason this test fails (IIRC) is that the real size in bytes of a tensor in spyre doesn't always match the host size, which causes some confusion. For example, a torch.int64 tensor will be represented in int32 in spyre, plus some padding for sticks usually, so a lot of asserts checking if the size is the expected one in the device fail.

**Comment by jason-liu227** [comment]
@prasannabiswas-iitb @ani300 
When working on this PR, I did come across this issue, at the time I couldn't find a reason for the error. Therefore I changed the approach and used a different way of checking. The idea is to be strict on comparing actions and somewhat loosely on sizes. 

In the PR, I have made the changes:
``` 
 else:
            for (act_action, act_size), (exp_action, exp_size) in zip(actual, expected):
                self.assertEqual(act_action, exp_action)
                self.assertGreaterEqual(
                    act_size, exp_size, f"Expected at least {exp_size}, got {act_size}"
                )
```

The "check" is split into 3 parts: 

1. the number of events
2. the exact action type 
3. the allocation size

Since there will always be padding, a loose check on size allows specific details on backend from failing.

Please confirm if this method is acceptable.

File: tests/profiler/test_spyre_profiler.py [hunk 1]
```diff
@@ -221,3 +221,66 @@ def test_kineto_memcpy_and_memset_events_captured():
 
     memset_events = [e for e in events if e.get("cat") == "gpu_memset"]
     assert memset_events, "Expected at least one memset event in the kineto-spyre trace"
+
+
+class TestMemoryProfilerTimeline(TestCase):
+    @unittest.skipIf(not Test_spyre, "spyre device required")
+    def test_memory_timeline_no_id(self, device="spyre") -> None:
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: This test always receives its default device="spyre"; moreover, the class-level Spyre skip prevents execution entirely without Spyre. Therefore Lines 275-280 are unreachable and the required CPU coverage is absent. Split CPU and Spyre into independently runnable tests (or parameterize them).

File: tests/profiler/test_spyre_profiler.py [hunk 2]
```diff
@@ -221,3 +221,66 @@ def test_kineto_memcpy_and_memset_events_captured():
...
+            expected = expected[2:]
+            for event in expected:
+                self.assertTrue(
+                    event in actual, f"event: {event} was not found in actual."
+                )
+        else:
+            for (act_action, act_size), (exp_action, exp_size) in zip(actual, expected):
+                self.assertEqual(act_action, exp_action)
+                self.assertGreaterEqual(
+                    act_size, exp_size, f"Expected at least {exp_size}, got {act_size}"
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: zip(actual, expected) truncates to the shorter list, so an actual timeline missing one or more trailing expected events can pass. Assert that actual has at least len(expected) entries before comparing the ordered prefix.


```suggestion
            self.assertGreaterEqual(len(actual), len(expected))
            for (act_action, act_size), (exp_action, exp_size) in zip(actual, expected):
                self.assertEqual(act_action, exp_action)
                self.assertGreaterEqual(
                    act_size, exp_size, f"Expected at least {exp_size}, got {act_size}"
```

File: tests/profiler/test_spyre_profiler.py (hunk 1 above)
Comment (kaoutar55) [inline tests/profiler/test_spyre_profiler.py]: The `device="spyre"` default combined with the class-level `@unittest.skipIf(not Test_spyre, ...)` means the `if device == "cpu":` branch at line 275 is never exercised. Remove the dead branch, or split into two test methods (one per device).

File: tests/profiler/test_spyre_profiler.py [hunk 3]
```diff
@@ -221,3 +221,66 @@ def test_kineto_memcpy_and_memset_events_captured():
...
+
+        # See above.
+        if device == "cpu":
+            expected = expected[2:]
+            for event in expected:
+                self.assertTrue(
+                    event in actual, f"event: {event} was not found in actual."
+                )
+        else:
+            for (act_action, act_size), (exp_action, exp_size) in zip(actual, expected):
```
Comment (kaoutar55) [inline tests/profiler/test_spyre_profiler.py]: `zip(actual, expected)` silently drops trailing `expected` entries when `actual` is shorter. Add a length guard before the loop:
```python
self.assertGreaterEqual(len(actual), len(expected), "actual timeline has fewer events than expected")
```

File: tests/profiler/test_spyre_profiler.py [hunk 4]
```diff
@@ -221,3 +221,66 @@ def test_kineto_memcpy_and_memset_events_captured():
...
+                self.assertEqual(act_action, exp_action)
+                self.assertGreaterEqual(
```
Comment (kaoutar55) [inline tests/profiler/test_spyre_profiler.py]: The `assertGreaterEqual` on size handles stick padding, but consider bounding the upper side as well. Without it, a 10x overallocation passes silently. Something like `self.assertLessEqual(act_size, exp_size * 4, ...)` would still accommodate alignment rounding while catching real bugs.

**Review by kaoutar55 — COMMENTED** [review]
@prasannabiswas-iitb, using `assertGreaterEqual` on sizes is a reasonable way to handle the stick-padding and int64→32 downcasting mismatch. Three items still need attention before this can merge:

1. The `if device == "cpu":` branch (lines 275-280) is unreachable. The method signature hardcodes `device="spyre"`, and the class-level skip prevents execution without hardware. Either remove the dead path or parameterize the test to run on both devices.
2. `zip(actual, expected)` silently truncates to the shorter list. If `actual` has fewer entries than `expected`, missing events will not fail the test. Add `self.assertGreaterEqual(len(actual), len(expected))` before the loop (per Vikram's suggestion).
3. The `>=` check on sizes is unbound on the upper side. A 10x or 100x overallocation would still pass. Adding an upper-bound assertion (e.g., `self.assertLessEqual(act_size, exp_size * N)` for some reasonable multiplier) would catch genuine allocation bugs that the current check masks.

The overall test logic is sound once those three are resolved.

**Comment by prasannabiswas-iitb** [comment]
Hi @jason-liu227, the PR looks good to me. 

Just a comment:

Why specifically 4x? Is this tied to known stick-padding behavior (e.g., 128-byte alignment on a 32-byte element)? it should be documented (e.g: a small comment)

So that 4x is a known thing, it is not supposed to mask real overallocation or cause flaky failures on future hardware.

**Comment by jason-liu227** [comment]
@prasannabiswas-iitb 
The 4x was originally part of @kaoutar55 suggestion above: 

"The assertGreaterEqual on size handles stick padding, but consider bounding the upper side as well. Without it, a 10x overallocation passes silently. Something like self.assertLessEqual(act_size, exp_size * 4, ...) would still accommodate alignment rounding while catching real bugs."

Initially I had questioned the 4x and wondered why it couldnt be other values (* N) or otherwise what defined a correct  * N.

I considered 2x but still strict and possibility of failing legitimate allocator changes.
On the contrary 8x would less likely to be flaky, but provides weaker protection.

**Comment by prasannabiswas-iitb** [comment]
> @prasannabiswas-iitb The 4x was originally part of @kaoutar55 suggestion above:
> 
> "The assertGreaterEqual on size handles stick padding, but consider bounding the upper side as well. Without it, a 10x overallocation passes silently. Something like self.assertLessEqual(act_size, exp_size * 4, ...) would still accommodate alignment rounding while catching real bugs."
> 
> Initially I had questioned the 4x and wondered why it couldnt be other values (* N) or otherwise what defined a correct * N.
> 
> I considered 2x but still strict and possibility of failing legitimate allocator changes. On the contrary 8x would less likely to be flaky, but provides weaker protection.

I see, this is completely fine. Just add a comment above the `*4` code snippet. Because if someone works on this test in future `*4` seems confusing as why we have some hardcoded value.  Add a comment "reasoning for 4x". Rest PR looks good to me.

File: tests/profiler/test_spyre_profiler.py [hunk 5]
```diff
@@ -221,3 +221,105 @@ def test_kineto_memcpy_and_memset_events_captured():
...
+        self.assertGreaterEqual(len(actual), len(expected))
+
+        for (act_action, act_size), (exp_action, exp_size) in zip(actual, expected):
+            self.assertEqual(act_action, exp_action)
+            self.assertGreaterEqual(
+                act_size, exp_size, f"Expected at least {exp_size}, got {act_size}"
+            )
+            self.assertLessEqual(
+                act_size,
+                exp_size * 4,
```
Comment (prasannabiswas-iitb) [inline tests/profiler/test_spyre_profiler.py:286]: Need a reasoning for 4x. Rest LGTM.

**Review by prasannabiswas-iitb — APPROVED** [review]

**Review by ppnaik1890 — APPROVED** [review]
lgtm. 
@anubhavjana can this be merged. Thanks!

**Review by anubhavjana — APPROVED** [review]
